File: testsuite/force_cap.py
#
# Copyright (C) 2017 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Tests particle property setters/getters
from __future__ import print_function
import unittest as ut
import espressomd
import numpy as np

@ut.skipIf(not espressomd.has_features("LENNARD_JONES"),
           "Skipped because of LENNARD_JONES ")
class ForceCap(ut.TestCase):
    """Tests the force capping mechanism.

    """

    s = espressomd.System()
    s.cell_system.skin = 0.0
    s.seed = range(s.cell_system.get_state()["n_nodes"])

    @classmethod
    def setUpClass(cls):
        np.random.seed(42)

    def calc_f_max(self):
        f = np.power(self.s.part[:].f, 2)
        sqr_sum = (np.sum(f, axis=1))
        f_max = np.max(sqr_sum)**0.5
        return f_max

    def test(self):
        N = 200
        f_cap = 10.
        s = self.s
        s.part.clear()
        s.time_step = 0.1
        s.box_l = 3*[10.]
        s.part.add(pos=10. * np.random.random((N, 3)))

        self.s.non_bonded_inter[0, 0].lennard_jones.set_params(
            epsilon=1000., sigma=2.,
            cutoff=1.5, shift=0.0)

        self.s.integrator.run(0)
        # Check that there is sth to cap
        self.assertGreater(self.calc_f_max(), f_cap)

        self.s.force_cap = f_cap

        # Check interface
        self.assertEqual(self.s.force_cap, f_cap)
        self.s.integrator.run(0)

        # Since there was a force larger than f_cap, the
        # maximum should now be f_cap.
        self.assertAlmostEqual(self.calc_f_max(), f_cap, places=7)

if __name__ == "__main__":
    print("Features: ", espressomd.features())
    ut.main()

File: src/core/forcecap.cpp
/*
 * Copyright (C) 2010-2019 The ESPResSo project
 * Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010
 *   Max-Planck-Institute for Polymer Research, Theory Group
 *
 * This file is part of ESPResSo.
 *
 * ESPResSo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ESPResSo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
/** \file
 *  Force capping calculation.
 */

#include "forcecap.hpp"

#include "Particle.hpp"
#include "communication.hpp"
#include "event.hpp"

#include <utils/Vector.hpp>
#include <utils/math/sqr.hpp>

#include <cmath>

static double force_cap = 0.0;

double forcecap_get() { return force_cap; }

void forcecap_cap(ParticleRange const &particles) {
  if (force_cap <= 0) {
    return;
  }

  auto const force_cap_sq = Utils::sqr(force_cap);

  for (auto &p : particles) {
    auto const force_sq = p.force().norm2();
    if (force_sq > force_cap_sq) {
      p.force() *= force_cap / std::sqrt(force_sq);
    }
  }
}

void mpi_set_forcecap_local(double force_cap) {
  ::force_cap = force_cap;
  on_forcecap_change();
}

REGISTER_CALLBACK(mpi_set_forcecap_local)

void mpi_set_forcecap(double force_cap) {
  mpi_call_all(mpi_set_forcecap_local, force_cap);
}
